File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.js
var maxPathSum = function(root) {
    let ans = -Infinity;

    function dfs(node) {
        if (!node) return 0;
        let left = Math.max(dfs(node.left), 0);
        let right = Math.max(dfs(node.right), 0);
        ans = Math.max(ans, node.val + left + right);
        return node.val + Math.max(left, right);
    }

    dfs(root);
    return ans;
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.c
int record;

int compare(int a, int b) {
    return a >= b ? a : b;
}

int recursive(struct TreeNode* node) {
    if (!node) return -1001;
    if (!node->left && !node->right) {
        record = record > node->val ? record : node->val;
        return node->val;
    } else {
        int leftRes = compare(recursive(node->left), 0);
        int rightRes = compare(recursive(node->right), 0);
        int val = node->val;
        int highest = compare(val, compare(val + leftRes, compare(val + rightRes, val + leftRes + rightRes)));
        record = record > highest ? record : highest;
        return compare(val, compare(val + leftRes, val + rightRes));
    }
}

int maxPathSum(struct TreeNode* root) {
    record = root->val;
    recursive(root);
    return record;
}